### PR TITLE
[PIPE-1079] Add maze runner tests for Spring notifier

### PR DESCRIPTION
**Pull request by Pezzah**
## Goal

Add maze runner tests to the Java notifier

## Design

Initial set of tests are a subset of the android notifier tests

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/EmptyStacktraceScenario.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Attempts to deliver a handled exception with no stacktrace.
+ */
+public class EmptyStacktraceScenario extends Scenario {
+    @Override
+    public void run() {
+        bugsnag.notify(new EmptyException("EmptyStacktraceScenario"));
+    }
+
+    private class EmptyException extends Throwable {
```
Comment (fractalwrench) [inline mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/EmptyStacktraceScenario.java]: this class could be static to avoid an implicit reference to the enclosing class

Comment (fractalwrench) [reply mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/EmptyStacktraceScenario.java]: the class is missing a serialVersionUid field

File: mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/ManualSessionScenario.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    @Override
+    public void run() {
+        bugsnag.setAppVersion("1.2.3");
+        bugsnag.startSession();
+
+        try {
+            Thread.sleep(1000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
```
Comment (fractalwrench) [inline mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/ManualSessionScenario.java]: Why is this delay necessary? This is not required in the android scenario.

File: mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+        ctx = applicationContext;
+    }
+
+    @Override
+    public void run(String... args) throws Exception {
+        // Put args into the system property so that they can be used later
+        for (String arg : args) {
+            String[] argParts = arg.split("=");
+            LOGGER.info("Setting property " + argParts[0] + "=" + argParts[1]);
```
Comment (fractalwrench) [inline mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java]: it'd be good to check that there are only 2 parts here, and throw a runtimeexception if that's not the case

File: mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    private static Scenario testCaseForName(String eventType) {
+
+        try {
+            Class clz = Class.forName("com.bugsnag.mazerunner.scenarios." + eventType);
+            Constructor constructor = clz.getConstructors()[0];
+            return (Scenario) constructor.newInstance();
+        } catch(Exception ex) {
+            return null;
```
Comment (fractalwrench) [inline mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java]: We may just want to throw the exception here, as something will have gone very wrong with the scenario setup.

File: .travis.yml [hunk 5]
```diff
@@ -13,16 +13,26 @@ jobs:
...
-    script: ./gradlew check :test
+    script:
+      - ./gradlew check :test
+      - bundle install
+      - bundle exec bugsnag-maze-runner
   - stage: test
     jdk: openjdk7
-    script: ./gradlew :test
+    script:
+      - ./gradlew check :test
```
Comment (fractalwrench) [inline .travis.yml]: this stage appears to be missing the mazerunner step

File: .travis.yml [hunk 6]
```diff
@@ -13,16 +13,26 @@ jobs:
   include:
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply .travis.yml]: I can't seem to get this to work right now, we should revisit this later

File: .travis.yml [hunk 7]
```diff
@@ -13,16 +13,26 @@ jobs:
...
   - stage: test
     jdk: openjdk8
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply .travis.yml]: moved

File: features/app_version.feature [hunk 8]
```diff
@@ -0,0 +1,9 @@
+Feature: Reporting app version
+
+Scenario: Test handled Android Exception
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/app_version.feature]: removed

File: features/ignored_reports.feature [hunk 9]
```diff
@@ -0,0 +1,9 @@
+Feature: Reports are ignored
+
+Scenario: Exception classname ignored
+    When I run "IgnoredExceptionScenario" with the defaults
+    Then I should receive no requests
+
+Scenario: Disabled Exception Handler
+    When I run "DisableAutoNotifyScenario" with the defaults
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/ignored_reports.feature]: Added the file

File: mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        // Create and run the test case
+        LOGGER.info("Creating test case");
+        Scenario s = testCaseForName(System.getProperty("EVENT_TYPE"));
+        if (s != null) {
+            LOGGER.info("running test case");
+            s.run();
+        } else {
+            LOGGER.info("No test case found!");
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java]: This is not really possible at the moment, because the spring application won't end on an unhandled exception. There is some code to explicitly close the application after running the test

File: mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/EmptyStacktraceScenario.java (hunk 1 above)
Comment (Pezzah) [reply mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/EmptyStacktraceScenario.java]: added

File: mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/ManualSessionScenario.java (hunk 2 above)
Comment (Pezzah) [reply mazerunner/src/main/java/com/bugsnag/mazerunner/scenarios/ManualSessionScenario.java]: removed

File: mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java (hunk 3 above)
Comment (Pezzah) [reply mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java]: added check

File: mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java (hunk 4 above)
Comment (Pezzah) [reply mazerunner/src/main/java/com/bugsnag/mazerunner/TestCaseRunnner.java]: for now this will be handled by the calling class, we need to handle what is going on to shutdown cleanly

File: .travis.yml (hunk 5 above)
Comment (Pezzah) [reply .travis.yml]: I removed this because it failed here https://travis-ci.org/bugsnag/bugsnag-java/builds/422924616?utm_source=github_status&utm_medium=notification I'll try adding it again
